clambcc/clambc-compiler.py

```diff
@@ -564,18 +623,118 @@ def fixFileSize(optimizedFile) :
+    for i in range(0, 10):
+        ret.append("%s%d%d" %(prog, maj, 9-i))
 
-def verifyClangInstallation():
-    exes = ["clang", "opt"]
-    for exe in exes:
+    return ret
+
+
+def findProgram(progName, clangVersion, strictVersion):
+    ret = None
+
+    clangList = []
+    clangList.extend(genList(progName, clangVersion))
+
+    if not strictVersion:
+        for i in CLANG_LLVM_KNOWN_VERSIONS:
+            if i == clangVersion:
+                continue
+            clangList.extend(genList(progName, i))
+
+    for c in clangList:
         try:
-            exe = f"{exe}-{CLANG_VERSION}"
-            subprocess.run([exe, "-v"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            subprocess.run([c, "-v"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            return c
         except FileNotFoundError:
-            print (f"{exe} must be installed and in your path.", file=sys.stderr)
-            return 1
+            pass
 
-    return 0
+    version = findVersion(progName)
+    if  version == clangVersion:
+        return progName
+
+    if not strictVersion:
+        if version >= MIN_CLANG_LLVM_VERSION:
+            return progName
+    else:
+        print(f"error: {progName} requested version not found", file=sys.stderr)
+
+
+
+    return ret
+
+
+def findClangLLVM(options):
+    cl = None
+
+    if (((None == options.clangBinary) and (None != options.optBinary)) 
+            or ((None != options.clangBinary) and (None == options.optBinary))):
+        print(f"error: if '{CLANG_BINARY_ARG} is present, then {OPT_BINARY_ARG} must also be present", out=sys.stderr)
+        sys.exit(1)
+
+
+    if None != options.clangBinary:
+        cl = ClangLLVM(options.clangBinary, options.optBinary)
+    else:
+
+        clangVersion = options.clangVersion
+        llvmVersion = options.llvmVersion
+        strictClang = True
+        strictLLVM = True
+
+        if None == clangVersion:
+            clangVersion = PREFERRED_CLANG_LLVM_VERSION
+            strictClang = False
+        else:
+            if not clangVersion.isnumeric():
+                print (f"error: {CLANG_VERSION_ARG} must be passed an integer type")
+                sys.exit(1)
+            clangVersion = int(clangVersion)
+
+
+        if None == options.llvmVersion:
+            llvmVersion = PREFERRED_CLANG_LLVM_VERSION
+            strictLLVM = False
+
+        else:
+            if not llvmVersion.isnumeric():
+                print (f"error: {LLVM_VERSION_ARG} must be passed an integer type")
+                sys.exit(1)
+            llvmVersion = int(llvmVersion)
+
+
+        clang = findProgram(CLANG_NAME, clangVersion, strictClang)
+        if None == clang:
+            print (f"{CLANG_NAME} must be installed and in your path.", file=sys.stderr)
+
+        opt = findProgram(LLVM_NAME, llvmVersion, strictLLVM)
+        if None == opt:
+            print (f"{LLVM_NAME} must be installed and in your path.", file=sys.stderr)
```

Comment:
no need for spaces between `print` and `(` (here and above).  Print is a function in python3.